### ISSUE TITLE
QR factorization unstable on GPU

### ISSUE THREAD

**Issue by lucasb-eyer** [tensorflow/tensorflow#20099]
### Describe the problem
Performing QR factorization of relatively large matrices on the GPU is numerically unstable. I am unsure if it's a tensorflow issue, or an upstrea cuSOLVE issue, as I don't have the time to write a custom cuSOLVE C call to test. The issue is neither present on the CPU, nor in numpy.

Steps to reproduce:

```
import tensorflow as tf
tf.enable_eager_execution()

with tf.device('/gpu:0'):
    q, r = tf.qr(tf.random_normal((2048, 1024))
    print(tf.norm(q, axis=0))
    print(tf.reduce_sum(tf.cast(tf.abs(tf.norm(q, axis=0) - 1) > 1, tf.float32)))
```

Run this multiple times and you get *wildly* different results, sometimes all norms are almost 1.0, sometimes the last entries have "slightly" larger norms 1-10, sometimes the last norms are huge (example: 3.1662303e+11), and sometimes they are NaN.

Running in float64 does not solve the issue.

Running on the CPU does solve the issue.

### System information

Tensorflow version as per `print(tf.GIT_VERSION, tf.VERSION)`:  `v1.7.0-3-g024aecf414 1.7.0` 

What follows is the output of `tf_env_collect.sh`:

Note 1: I cannot upgrade to latest TF to check it there right now.
Note 2: Multiple cuda versions are shown below, but only 9.0 is the first one in all `*PATH` variables.

```
== cat /etc/issue ===============================================
Linux grimbergen 4.4.0-116-generic #140~14.04.1-Ubuntu SMP Fri Feb 16 09:25:20 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
VERSION="14.04.5 LTS, Trusty Tahr"
VERSION_ID="14.04"

== are we in docker =============================================
No

== compiler =====================================================
c++ (Ubuntu 4.8.4-2ubuntu1~14.04.4) 4.8.4
Copyright (C) 2013 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


== uname -a =====================================================
Linux grimbergen 4.4.0-116-generic #140~14.04.1-Ubuntu SMP Fri Feb 16 09:25:20 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

== check pips ===================================================
numpy              1.14.2      
protobuf           3.5.2.post1 
tensorflow-gpu     1.7.0       

== check for virtualenv =========================================
False

== tensorflow import ============================================
tf.VERSION = 1.7.0
tf.GIT_VERSION = v1.7.0-3-g024aecf414
tf.COMPILER_VERSION = v1.7.0-3-g024aecf414
/home/beyer/.pyenv/versions/tensorflow-latest/lib/python3.6/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters

== env ==========================================================
LD_LIBRARY_PATH /home/beyer/inst/cudnn7/lib64:/usr/local/cuda-9.0/extras/CUPTI/lib64/:/usr/local/cuda-9.0/lib64:/home/beyer/inst/cudnn6/lib64:/usr/local/cuda-8.0/extras/CUPTI/lib64/:/usr/local/cuda-8.0/lib64:/home/beyer/inst/lib64:/home/beyer/inst/lib:
DYLD_LIBRARY_PATH is unset

== nvidia-smi ===================================================
Mon Jun 18 18:14:33 2018       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 384.111                Driver Version: 384.111                   |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  TITAN X (Pascal)    Off  | 00000000:01:00.0 Off |                  N/A |
| 30%   50C    P8    22W / 250W |  11591MiB / 12187MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   1  GeForce GTX 108...  Off  | 00000000:02:00.0 Off |                  N/A |
| 44%   39C    P2    60W / 250W |  10625MiB / 11172MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0      4713      C   ...3.6.2/envs/tensorflow-latest/bin/python 11581MiB |
|    1      4713      C   ...3.6.2/envs/tensorflow-latest/bin/python 10615MiB |
+-----------------------------------------------------------------------------+

== cuda libs  ===================================================
/usr/local/cuda-8.0/doc/man/man7/libcudart.7
/usr/local/cuda-8.0/doc/man/man7/libcudart.so.7
/usr/local/cuda-8.0/lib64/libcudart_static.a
/usr/local/cuda-8.0/lib64/libcudart.so.8.0.44
/usr/local/cuda-9.0/doc/man/man7/libcudart.7
/usr/local/cuda-9.0/doc/man/man7/libcudart.so.7
/usr/local/cuda-9.0/lib64/libcudart_static.a
/usr/local/cuda-9.0/lib64/libcudart.so.9.0.176
/usr/local/cuda-7.5/lib/libcudart_static.a
/usr/local/cuda-7.5/lib/libcudart.so.7.5.18
/usr/local/cuda-7.5/doc/man/man7/libcudart.7
/usr/local/cuda-7.5/doc/man/man7/libcudart.so.7
/usr/local/cuda-7.5/lib64/libcudart_static.a
/usr/local/cuda-7.5/lib64/libcudart.so.7.5.18

**Comment by asimshankar** [comment]
@chsigg @zheng-xq - Mind taking a look?

**Comment by lucasb-eyer** [comment]
Just adding that I found this out when orthogonal_initializer for a large matrix sometimes blew up my runs from the start and sometimes didn't.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @chsigg: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @chsigg: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @chsigg: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @chsigg: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @chsigg: It has been 74 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @chsigg: It has been 89 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @chsigg: It has been 104 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by rajatmonga** [comment]
@rmlarsen in case you have some ideas.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @chsigg: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @chsigg: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by azaks2** [comment]
There was/is a bug in cuSOLVE, so the gpu kernel now calls cpu one (to avoid collocation problems with variable initialization that uses qr).
Take a look at https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/qr_op_float.cc. if interested.

**Comment by albertz** [comment]
Hi @lucasb-eyer, somewhat funny that we stumble upon the same problems. :) (Actually via @vieting.)

**Comment by lucasb-eyer** [comment]
Haha hey @albertz ! Small world indeed, hope to also bump into you again at conferences when things go back to normal =)
